Kudu.Core/Functions/FunctionManager.cs

```diff
@@ -49,52 +51,214 @@ public FunctionManager(IEnvironment environment, ITraceFactory traceFactory)
                         {
-                            tracer.Trace(String.Format("Skip {0} of {1}", type, functionJson));
+                            tracer.Trace(String.Format("Skip {0} of {1}", type, functionJson.Name));
                         }
                     }
                 }
                 catch (Exception ex)
                 {
-                    tracer.Trace(String.Format("{0} is invalid. {1}", functionJson, ex.Message));
+                    tracer.Trace(String.Format("{0} is invalid. {1}", functionJson.Name, ex.Message));
                 }
             }
 
             return inputs;
         }
 
-        public IEnumerable<string> EnumerateFunctionFiles()
+        public async Task<FunctionEnvelope> CreateOrUpdateAsync(string name, FunctionEnvelope functionEnvelope)
+        {
+            var functionDir = Path.Combine(_environment.FunctionsPath, name);
+
+            // Make sure the function folder exists
+            FileSystemHelpers.EnsureDirectory(functionDir);
+
+            // If files are included, write them out
+            if (functionEnvelope?.Files != null)
+            {
+                // Delete all existing files in the directory. This will also delete current function.json, but it gets recreated below
+                FileSystemHelpers.DeleteDirectoryContentsSafe(functionDir);
+
+                foreach (var fileEntry in functionEnvelope?.Files)
+                {
+                    await FileSystemHelpers.WriteAllTextToFileAsync(Path.Combine(functionDir, fileEntry.Key), fileEntry.Value);
+                }
+            }
+
+            // Create the function.json
+            await FileSystemHelpers.WriteAllTextToFileAsync(Path.Combine(functionDir, Constants.FunctionsConfigFile), JsonConvert.SerializeObject(functionEnvelope?.Config ?? new JObject()));
+            return await GetFunctionConfigAsync(name);
+        }
+
+        public async Task<IEnumerable<FunctionEnvelope>> ListFunctionsConfigAsync()
+        {
+            var configList = await Task.WhenAll(
+                    FileSystemHelpers
+                    .GetDirectories(_environment.FunctionsPath)
+                    .Select(d => Path.Combine(d, Constants.FunctionsConfigFile))
+                    .Where(FileSystemHelpers.FileExists)
+                    .Select(async f => { try { return await GetFunctionConfigAsync(Path.GetFileName(Path.GetDirectoryName(f))); } catch { return null; } }));
+            return configList.Where(c => c != null);
+        }
+
+        public async Task<FunctionEnvelope> GetFunctionConfigAsync(string name)
+        {
+            var path = Path.Combine(GetFunctionPath(name), Constants.FunctionsConfigFile);
+            if (FileSystemHelpers.FileExists(path))
+            {
+                return CreateFunctionConfig(await FileSystemHelpers.ReadAllTextFromFileAsync(path), name);
+            }
+
+            throw new FileNotFoundException($"Function ({path}) does not exist");
```

Comment:
Also, it's more correct to say that it's the function *config* that doesn't exist. If the function itself doesn't `GetFunctionPath` above will throw.